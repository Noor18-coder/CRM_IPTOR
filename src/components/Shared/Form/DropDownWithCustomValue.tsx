/* eslint-disable jsx-a11y/control-has-associated-label */
import React from 'react';
import * as models from '../../../helpers/Api/models';

interface SelectProps {
  obj: models.AttributeFormField[];
  options: any;
  onSelect: (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => void;
  addElement?: (data: models.AttributeFormField) => void;
  deleteElement?: (data: models.AttributeFormField) => void;
  error?: string;
}

const DropDownWithCustomValue: React.FC<SelectProps> = ({ obj, options, onSelect, addElement, deleteElement }) => {
  const attributeValues = options
    ? options.data.find((dropdownObj: models.DropDownValues) => {
        return dropdownObj.attributeId === obj[0].attributeId;
      })
    : null;

  return (
    <>
      <div className="adding-txt-field-container">
        {obj[0].uniqueRecord === false && addElement ? (
          <button type="button" className="btn btn-link add-txt-field" onClick={() => addElement(obj[0])}>
            + ADD
          </button>
        ) : null}
        {obj.map((fieldObj: models.AttributeFormField, index: number) => {
          if (index === 0) {
            return (
              <div className="form-group oppty-form-elements">
                <label htmlFor={obj[0].attributeType} className="opp-label">
                  {fieldObj.description}
                </label>
                <div>
                  <input
                    type="text"
                    className="form-control"
                    placeholder={`${obj[0].description}`}
                    id={fieldObj.id}
                    value={fieldObj.attributeValue}
                    onChange={onSelect}
                    list={`OPTIONS_${fieldObj.id}`}
                  />
                  <datalist id={`OPTIONS_${fieldObj.id}`}>
                    {attributeValues?.values.map((valueObj: models.DropDownValue) => {
                      const optionSelected = valueObj.valueField === obj[0].attributeValue;
                      return valueObj.fieldDescription ? (
                        <option value={valueObj.valueField} selected={optionSelected}>
                          {valueObj.valueField} - {valueObj.fieldDescription}
                        </option>
                      ) : (
                        <option value={valueObj.valueField} selected={optionSelected}>
                          {valueObj.valueField}
                        </option>
                      );
                    })}
                  </datalist>
                </div>
              </div>
            );
          } else {
            return (
              <div className="form-element-with-delete-button">
                <div className="form-group oppty-form-elements add-new-ipfield">
                  <input
                    type="text"
                    className="form-control"
                    placeholder={`${obj[0].description}`}
                    id={fieldObj.id}
                    value={fieldObj.attributeValue}
                    onChange={onSelect}
                    list={`OPTIONS_${fieldObj.id}`}
                  />
                  <datalist id={`OPTIONS_${fieldObj.id}`}>
                    {attributeValues?.values.map((valueObj: models.DropDownValue) => {
                      const optionSelected = valueObj.valueField === obj[0].attributeValue;
                      return valueObj.fieldDescription ? (
                        <option value={valueObj.valueField} selected={optionSelected}>
                          {valueObj.valueField} - {valueObj.fieldDescription}
                        </option>
                      ) : (
                        <option value={valueObj.valueField} selected={optionSelected}>
                          {valueObj.valueField}
                        </option>
                      );
                    })}
                  </datalist>
                </div>
                {/* onClick={() => deleteElement(fieldObj)} */}
                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
                {deleteElement && <button type="button" className="btn btn-link ip-delete-field" onClick={() => deleteElement(fieldObj)} />}
              </div>
            );
          }
        })}
      </div>
    </>
  );
};

export default DropDownWithCustomValue;
