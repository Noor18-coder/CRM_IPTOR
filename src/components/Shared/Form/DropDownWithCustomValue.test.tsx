import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import DropDownWithCustomValue from './DropDownWithCustomValue';
import StoreMock from '../../../mocks/Store.mock';

const store = StoreMock.createInitialStore();

const obj = [
  {
    parentFile: 'jhj',
    attributeType: 'jshjd',
    description: 'ndndss',
    preloaded: true,
    allowCopy: true,
    valueFormat: 'jdsh',
    valueFormatDesc: 'dfhvf',
    attributeId: 'dghgd',
    validate: true,
    uniqueRecord: true,
    statement: 'hdsghdsg',
    sequence: 4736,
    group: 'sfhsdg',
    initializeAttribute: true,
    initialValueN: 867,
    protectData: true,
    valuesExist: true,
  },
  {
    parentFile: 'jjfg',
    attributeType: 'wrwe',
    description: 'lkjl',
    preloaded: true,
    allowCopy: true,
    valueFormat: 'vnv',
    valueFormatDesc: 'iui',
    attributeId: 'dasd',
    validate: true,
    uniqueRecord: true,
    statement: 'uoyu',
    sequence: 4736,
    group: 'xvss',
    initializeAttribute: true,
    initialValueN: 867,
    protectData: true,
    valuesExist: true,
  },
];
const options = {
  data: [
    { values: 'Option1', attributeId: 'id1' },
    { values: 'Option2', attributeId: 'id2' },
    { values: 'Option3', attributeId: 'id3' },
  ],
};
const onSelect = jest.fn();

describe('[Shared] DropDownWithCustomValue', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DropDownWithCustomValue obj={obj} options={options} onSelect={onSelect} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
