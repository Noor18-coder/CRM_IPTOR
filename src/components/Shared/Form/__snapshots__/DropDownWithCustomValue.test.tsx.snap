// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Shared] DropDownWithCustomValue should renders correctly 1`] = `
<div
  className="adding-txt-field-container"
>
  <div
    className="form-group oppty-form-elements"
  >
    <label
      className="opp-label"
      htmlFor="jshjd"
    >
      ndndss
    </label>
    <div>
      <input
        className="form-control"
        list="OPTIONS_undefined"
        onChange={[MockFunction]}
        placeholder="ndndss"
        type="text"
      />
      <datalist
        id="OPTIONS_undefined"
      />
    </div>
  </div>
  <div
    className="form-element-with-delete-button"
  >
    <div
      className="form-group oppty-form-elements add-new-ipfield"
    >
      <input
        className="form-control"
        list="OPTIONS_undefined"
        onChange={[MockFunction]}
        placeholder="ndndss"
        type="text"
      />
      <datalist
        id="OPTIONS_undefined"
      />
    </div>
  </div>
</div>
`;
